modules/assim.sequential/inst/sda.tobit.simulation.R

```diff
@@ -62,22 +64,213 @@ library(mvtnorm)
+  run = function(x = double(1), alpha = double(1), size = double(0), log = integer(0)){
+    returnType(double(0))
+    logProb <- lgamma(sum(alpha)) - sum(lgamma(alpha)) + sum(lgamma(alpha + x)) - lgamma(sum(alpha) + size)
+    
+    if(log) {
+      return(logProb)
+    } else {
+      return(exp(logProb))
+    }
+    
+  }
+)
+
+# set up the "r" function
+rdirchmulti <- nimbleFunction(
+  run = function(n = integer(0), alpha = double(1), size = double(0)) {
+    returnType(double(1))
+    if(n != 1) nimPrint("rdirchmulti only allows n = 1; using n = 1.")
+    p <- rdirch(1, alpha)
+    return(rmulti(1, size = size, prob = p))
+  })
+
+# tell NIMBLE about the newly available distribution
+registerDistributions(list(ddirchmulti = list(BUGSdist = "ddirchmulti(alpha, size)", 
+                                              types = c('value = double(1)', 'alpha = double(1)'))))
+
+dwtmnorm <- nimbleFunction(
+  run = function(x = double(1), mean = double(1), cov = double(2), wt = double(0), log = integer(0)){
+    returnType(double(0))
+    Prob <- dmnorm_chol(x, mean, chol(cov), prec_param = FALSE) * wt
+    return(Prob)
+  }
+)
+rwtmnorm <- nimbleFunction(
+  run = function(n = integer(0), mean = double(1), cov = double(2), wt = double(0)){
+    returnType(double(1))
+    if(n != 1) nimPrint("rdirchmulti only allows n = 1; using n = 1.")
+    Prob <- rmnorm_chol(n, mean, chol(cov), prec_param = FALSE) * wt
+    return(Prob)
+  }
+)
+
+dwtmnorm <- nimbleFunction(
+  run = function(x = double(1), mean = double(1), cov = double(2), wt = double(0), log = integer(0)){
+    returnType(double(0))
+    n <- nrow(cov)
+    logProb <- n * log(2 * pi) + log(det(cov)) + (x-mean) %*% inverse(cov) %*% t(x-mean) 
+    logProb <- -1/2 * wt * Prob
+    
+    if(log) return(logProb) else return(exp(logProb))
+  }
+)
+registerDistributions(list(dwtmnorm = list(BUGSdist = "dwtmnorm(mean, cov, wt)", 
+                                              types = c('value = double(1)','mean = double(1)', 'cov = double(2)', 'wt = double(0)'))))
+
+
+tobit2space.model <- nimbleCode({
+  
+  for(i in 1:N){
+    #row[i] ~ dcat(wts[1:N])
+    #s[i] ~ dcat(wts[1:N])
+    #d[i] <- myCalculation(grid[1:N], s[i])
+    y.censored[i,1:J] ~ dwtmnorm(mean = muf[1:J], cov = pf[1:J,1:J], wt = wts[i])
+    for(j in 1:J){
+      y.ind[i,j] ~ dinterval(y.censored[i,j], 0)
+    }
+  }
+  
+  #weighting the likelihood
+  #row[i] ~ dcat(weights)
+  #X[row[j],]
+  
+  muf[1:J] ~ dmnorm(mean = mu_0[1:J], cov = pf[1:J,1:J])
+  
+  Sigma[1:J,1:J] <- lambda_0[1:J,1:J]/nu_0
+  pf[1:J,1:J] ~ dinvwish(S = Sigma[1:J,1:J], df = J)
+  
+})
+
+x.ind <- x.censored <- matrix(NA, ncol=ncol(X), nrow=nrow(X))
+for(j in seq_along(mu.f)){
+  for(n in seq_len(nrow(X))){
+    x.ind[n,j] <- as.numeric(X[n,j] > 0)
+    x.censored[n,j] <- as.numeric(ifelse(X[n,j] > intervalX[j,2], 0, X[n,j])) #
+  }
+}
+
+
+wts <- matrix(NA,20,50)
+for(i in 1:20){
+  wts[i,] <- sample(size = 50,x=1:50,replace = F)
+}
+
+wts <- rep(1,50)
```

Comment:
again hardcoded values, might help assigning values to named objects even if they stay hardcoded